unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp

```diff
@@ -0,0 +1,231 @@
+      class_typet class_type =
+        require_symbol::require_complete_class(class_symbol);
+
+      java_class_typet java_class_type = to_java_class_type(class_type);
+      REQUIRE_FALSE(is_java_generics_class_type(java_class_type));
+
+      // TODO (tkiley): Extend this unit test when bounds are correctly
+      // parsed.
+#if 0
+      java_generics_class_typet java_generics_class_type=
+        to_java_generics_class_type(java_class_type);
+      REQUIRE(java_generics_class_type.generic_types().size()==1);
+      typet &type_var=java_generics_class_type.generic_types().front();
+      REQUIRE(is_java_generic_parameter(type_var));
+      java_generic_parametert generic_type_var=
+        to_java_generic_parameter(type_var);
+
+      REQUIRE(
+        generic_type_var.type_variable().get_identifier()==
+        doubleboundedinner_name+"::T");
+      REQUIRE(
+        java_generics_class_type_var(0, java_generics_class_type)==
+        doubleboundedinner_name+"::T");
+      THEN("Bound must be Number and Interface")
+      {
+
+      }
+#endif
+    }
+  }
+
+  GIVEN("An inner class with multiple generic parameters")
+  {
+    std::string twoelementinner_name = class_prefix + "$TwoElementInner";
+    REQUIRE(new_symbol_table.has_symbol(twoelementinner_name));
+    THEN("Both generic parameters should be encoded")
+    {
+      const symbolt &class_symbol =
+        new_symbol_table.lookup_ref(twoelementinner_name);
+      class_typet class_type =
+        require_symbol::require_complete_class(class_symbol);
+      java_generics_class_typet java_generics_class_type =
+        require_type::require_java_generic_class(class_type);
+
+      REQUIRE(java_generics_class_type.generic_types().size() == 2);
+
+      auto generic_param_iterator =
+        java_generics_class_type.generic_types().cbegin();
+
+      // The first parameter should be called K
+      {
+        const typet &first_param = *generic_param_iterator;
+        REQUIRE(is_java_generic_parameter(first_param));
+        java_generic_parametert generic_type_var =
+          to_java_generic_parameter(first_param);
+
+        REQUIRE(
+          generic_type_var.type_variable().get_identifier() ==
+          twoelementinner_name + "::K");
+        REQUIRE(
+          java_generics_class_type_var(0, java_generics_class_type) ==
+          twoelementinner_name + "::K");
+      }
+
+      ++generic_param_iterator;
+
+      // The second parameter should be called V
+      {
+        const typet &second_param = *generic_param_iterator;
+        REQUIRE(is_java_generic_parameter(second_param));
+        java_generic_parametert generic_type_var =
+          to_java_generic_parameter(second_param);
+
+        REQUIRE(
+          generic_type_var.type_variable().get_identifier() ==
+          twoelementinner_name + "::V");
+        REQUIRE(
+          java_generics_class_type_var(1, java_generics_class_type) ==
+          twoelementinner_name + "::V");
+      }
```

Comment:
Since the code for the first and second parameters is duplicate apart from the identifier name, it's worth thinking about whether this can be factored our into a helper function - probably the same one I previously mentioned :-)